File: Project/Backend/src/ArtAuction.API/Controllers/PostSoldController.cs
using ArtAuction.Application.DTOs.PostSold;
using ArtAuction.Application.Interfaces.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ArtAuction.API.Controllers;

[ApiController]
[Route("[controller]")]
public class PostSoldController : ControllerBase
{
    // Attributes
    private readonly IPostSoldServices _postSoldServices;
    
    // Constructor
    public PostSoldController(IPostSoldServices postSoldServices)
    {
        this._postSoldServices = postSoldServices;
    }
    
    // Get post sold API
    [Authorize(Roles = "Admin")]
    [HttpGet("GetAllPostSold")]
    public async Task<IActionResult> GetAllPostSold()
    {
        // Return all post sold
        return Ok(await _postSoldServices.GetAllPostSolds());
    }
    
    // Get Unpaid posts API
    [Authorize(Roles = "Buyer")]
    [HttpGet("GetUnpaidPostForBuyer")]
    public async Task<IActionResult> GetUnpaidPostForBuyer(int buyerId)
    {
        // Return unpaid post for buyer
        return Ok(await _postSoldServices.GetUnpaidPostSoldForBuyer(buyerId));
    }
    
    // Get buyer post bids
    [Authorize(Roles = "Buyer")]
    [HttpGet("GetPostSoldForBuyer")]
    public async Task<IActionResult> GetPostSoldForBuyer(int buyerId)
    {
        return Ok(await _postSoldServices.GetBuyerPostSolds(buyerId));
    }
    
    // Mark as paid API
    [Authorize(Roles = "Buyer")]
    [HttpPatch("MarkAsPaid")]
    public async Task<IActionResult> MarkAsPaid([FromBody] PostSoldPaidDto postSoldPaidDto)
    {
        // Check model state
        if (!ModelState.IsValid)
            return BadRequest(ModelState);
        
        // Check stats
        var result = await _postSoldServices.MarkAsPaid(postSoldPaidDto);
        if (!result)
            return BadRequest("Failed to mark the post as paid. Verify the Post and Buyer IDs.");
        
        // Return success stats
        return Ok("Successfully marked as paid.");
    }
}

File: AuctionServices/src/main/java/com/artauction/auctionservice/Controllers/PostSoldController.java
package com.artauction.auctionservice.Controllers;

import com.artauction.auctionservice.Aops.Roles.RequiresRole;
import com.artauction.auctionservice.Dtos.PostSold.PostSoldPaidDto;
import com.artauction.auctionservice.Services.Interfaces.PostSoldServices;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Arrays;

@RestController
@RequestMapping("/PostSold")
@RequiredArgsConstructor
public class PostSoldController {
    // Attributes
    private final PostSoldServices postSoldServices;

    // Get post sold API
    @RequiresRole("Admin")
    @GetMapping("/GetAllPostSold")
    public ResponseEntity<?> GetAllPostSold() throws Exception {
        return ResponseEntity.ok(postSoldServices.getAllPostSolds());
    }

    // Get unpaid posts API
    @RequiresRole("Buyer")
    @GetMapping("/GetUnpaidPostForBuyer")
    public ResponseEntity<?> GetUnpaidPostForBuyer(@RequestParam Integer buyerId) {
        return ResponseEntity.ok(postSoldServices.getUnpaidPostSoldForBuyer(buyerId));
    }

    // Get buyer post solds API
    @RequiresRole("Buyer")
    @GetMapping("/GetPostSoldForBuyer")
    public ResponseEntity<?> GetPostSoldForBuyer(@RequestParam Integer buyerId) {
        return ResponseEntity.ok(postSoldServices.getBuyerPostSolds(buyerId));
    }

    // Mark as paid API
    @RequiresRole("Buyer")
    @PatchMapping("/MarkAsPaid")
    public ResponseEntity<?> MarkAsPaid(@RequestBody PostSoldPaidDto postSoldPaidDto) {
        // Check result
        boolean result = postSoldServices.markAsPaid(postSoldPaidDto);
        if (!result)
            return ResponseEntity.badRequest()
                    .body("Failed to mark the post as paid. Verify the Post and Buyer IDs.");

        // Return success
        return ResponseEntity.ok("Successfully marked as paid.");
    }
}
